README.md
# mac-util

一个基于Mac上使用的工具命令行应用，可以帮助Mac用户快速地配置macOS系统。

这个CLI应用可以一键**开启任何来源**，或者安装`oh-my-zsh`以及与zsh相关的插件例如`zsh-autosuggestions`和`zsh-syntax-highlighting`，后续会支持更多的设置。


## 安装`mac-util`

全局安装`mac-util`：

```shell
yarn global add mac-util 
# 或者 npm install -g mac-util
```

现在你可以在命令行中直接输入`mac-util`。

## CLI 使用

### init

在命令行中输入：

```shell
mac-util init
```

之后会出现一系列选项：

```shell
➜ ~ mac-util init
? 选择你要进行的操作 (Use arrow keys)
❯ 1. 开启任何来源
  2. 关闭任何来源
  3. 安装oh-my-zsh
  4. 卸载oh-my-zsh
  5. NVM变量添加
  6. 安装zsh-autosuggestions插件
  7. 安装zsh-syntax-highlighting插件
```

你只需要通过方向键上下移动光标到需要设置的选项，然后按下回车键确认即可。CLI会自动帮你执行对应的操作。

例如我选中了`安装zsh-autosuggestions`，按下回车之后的执行效果如下：

```shell
➜  ~ mac-util init
? 选择你要进行的操作 6. 安装zsh-suggestions插件
开始安装 zsh-autosuggestions
运行命令：git clone https://github.91chi.fun/https://github.com/zsh-users/zsh-autosuggestions.git ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/zsh-autosuggestions
Cloning into '/Users/bigdog/.oh-my-zsh/custom/plugins/zsh-autosuggestions'...
安装zsh-autosuggestions成功...
更改~/.zshrc文件，写入zsh-autosuggestions...
完成
```

enjoy yourself!

## TODO

- [ ] 新增homebrew的自动安装

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/johnhom1024/mac-cli/compare/v1.0.0...v1.1.0) (2022-07-13)


### Features

* 去掉clear-console 移动@types/fs-extra到dev依赖中 ([58b3546](https://github.com/johnhom1024/mac-cli/commit/58b3546952a7140c0d11173db6da2ab4a0456b1a))
* 新增zsh-syntax-highlighting安装函数 ([5c02868](https://github.com/johnhom1024/mac-cli/commit/5c02868b69b98703584ab41a96bcad7dbc2a10a7))

## 1.0.0 (2022-07-13)


### Features

* 初始化项目 ([799b373](https://github.com/johnhom1024/mac-cli/commit/799b37385a054ad3b5065087afd652ebd51b6199))
* 调整error的类型合法性 ([755c6d1](https://github.com/johnhom1024/mac-cli/commit/755c6d13ce2016ce5c4ddda73fefced56447985d))
* 对命令函数进行封装 ([3e38d0e](https://github.com/johnhom1024/mac-cli/commit/3e38d0e388aaa834e9dbe006ee10e8a834ae673d))
* 给zshrc写入zsh-autosuggestions时判断plugins中是否已存在 ([b21f907](https://github.com/johnhom1024/mac-cli/commit/b21f9074f026aec761ea3d8dc7be39c006380d92))
* 更新.gitignore ([3332ed9](https://github.com/johnhom1024/mac-cli/commit/3332ed981c711624de37729e375e21409a0369b2))
* 更新package.json ([88d2596](https://github.com/johnhom1024/mac-cli/commit/88d2596331a751028bb36950099e4e9118f282de))
* 更新tsconfig.json ([246fe22](https://github.com/johnhom1024/mac-cli/commit/246fe2279e82da0502a31748761f8a5341461527))
* 更新yarn.lock ([f0e4d42](https://github.com/johnhom1024/mac-cli/commit/f0e4d421a638aa28c960408733dfdb8d117beef4))
* 删除lib/index.js ([c0c3658](https://github.com/johnhom1024/mac-cli/commit/c0c36583f6db09d864aab1eaaa5260bb1bc5d3e8))
* 完善目录结构 ([ce7ce08](https://github.com/johnhom1024/mac-cli/commit/ce7ce089d6059bc353710197b64212b225c4977d))
* 新增部分types包 ([e53d9bc](https://github.com/johnhom1024/mac-cli/commit/e53d9bcfd6045be392f78d5e34307b0aeea8e924))
* 新增卸载ohmyzsh命令 ([5594379](https://github.com/johnhom1024/mac-cli/commit/55943797729bd2acf75baa1a2f18a4cb7f39c75b))
* 新增cli.js的可执行类型 ([b9fd386](https://github.com/johnhom1024/mac-cli/commit/b9fd3866072d6275ae233f455af38f7304b5207a))
* 新增logger工具函数 ([5f9a7f6](https://github.com/johnhom1024/mac-cli/commit/5f9a7f630e8fdd7b739ad9d59583da81aee77921))
* 新增logger工具函数 ([4521e2a](https://github.com/johnhom1024/mac-cli/commit/4521e2a03d127fdb7387643ff7bd11ecfa567dd8))
* 新增README.md文件 ([8d4b570](https://github.com/johnhom1024/mac-cli/commit/8d4b570710535d97af29b90148b374eff4ce1bc2))
* 新增standard-vesion ([67a3810](https://github.com/johnhom1024/mac-cli/commit/67a381032e44e67f873fba6b3dbe6231f70341e4))
* 新增zsh-autosuggestions插件安装命令 ([b233f1d](https://github.com/johnhom1024/mac-cli/commit/b233f1d9e80533829deb8276317abc0e69c39b53))
* **init:** 新增init指令 ([a3b5775](https://github.com/johnhom1024/mac-cli/commit/a3b5775ced8bd0f2a299ef5ac425c065bce7d23c))
* **NVM:** 新增NVM变量配置 ([bddf1be](https://github.com/johnhom1024/mac-cli/commit/bddf1befd25ce30a96d23ff41d9d6efbd70eab8a))


src/order/init.ts
/*
 * @Date: 2022-03-07 17:51:05
 * @Author: handsome_anthony
 * @LastEditors: handsome_anthony
 * @Description: 初始化macOS的配置
 */

import { prompt } from 'inquirer';
import {
  openAnySource,
  closeAnySource,
  installOhmyzsh,
  uninstallOhmyzsh,
  addNVMToZshrc,
  installZshAutosuggestions,
  installZshSyntaxHighlighting,
} from '../utils/shellCommand';

async function init(): Promise<void> {
  const options = [
    {
      value: 1,
      name: '1. 开启任何来源'
    },
    {
      value: 2,
      name: '2. 关闭任何来源'
    },
    {
      value: 3,
      name: '3. 安装oh-my-zsh'
    },
    {
      value: 4,
      name: '4. 卸载oh-my-zsh'
    },
    {
      value: 5,
      name: '5. NVM变量添加'
    },
    {
      value: 6,
      name: '6. 安装zsh-autosuggestions插件'
    },
    {
      value: 7,
      name: '7. 安装zsh-syntax-highlighting插件'
    },
  ]
  const selectedOption = await prompt([
    {
      type: 'list',
      name: 'value',
      message: '选择你要进行的操作',
      choices: options,
    }
  ])

  try {
    switch (selectedOption.value) {
      case 1:
        openAnySource();
        break;
      case 2:
        closeAnySource();
        break;
      case 3:
        installOhmyzsh();
        break;
      case 4:
        uninstallOhmyzsh();
        break;
      case 5:
        addNVMToZshrc();
        break;
      case 6:
        installZshAutosuggestions();
        break;
      case 7:
        installZshSyntaxHighlighting();
        break;
      default:
        break;
    }
  } catch (error: unknown) {
    if (error instanceof Error) {
      console.error(error.message)
    }
  }

}



export default init;

src/utils/logger.ts
/*
 * @Author: johnhom1024 924358746@qq.com
 * @Date: 2022-06-22 20:44:49
 * @LastEditors: handsome_anthony
 * @LastEditTime: 2022-07-13 12:53:09
 * @FilePath: /mac-cli/src/utils/logger.ts
 * @Description: 这是默认设置,请设置`customMade`, 打开koroFileHeader查看配置 进行设置: https://github.com/OBKoro1/koro1FileHeader/wiki/%E9%85%8D%E7%BD%AE
 */
import chalk from 'chalk';

export function log(...args: any) {
  console.log(...args);
}

export function logInfo(msg: string) {
  console.log(
    chalk.bgBlue.black(' Info ') + ' ' +
    msg
  );
}

export function logSuccess(msg: string) {
  console.log(
    chalk.bgGreen.black(' Success ') + ' ' +
    msg
  );
}

export function logWarn(msg: string, logBlock = false) {
  if (logBlock) log('\n\n');

  console.log(
    chalk.bgYellow.black(' Warn ') + ' ' +
    chalk.yellow(msg)
  );

  if (logBlock) log('\n\n');
}

commitlint.config.js
/*
 * @Date: 2022-03-07 16:58:06
 * @Author: handsome_anthony
 * @LastEditors: handsome_anthony
 * @Description: 
 */

module.exports = {
    // 符合angular的校验规则
    extends: ["@commitlint/config-conventional"],
    
    rules: {
      "type-enum": [
        2,
        "always",
        [
          "feat",
          "fix",
          "build",
          "refactor",
          "style",
          "docs",
          "chore",
          "ci",
          "perf",
          "test",
        ],
      ],
      "type-case": [0],
      "type-empty": [0],
      "scope-empty": [0],
      "scope-case": [0],
      "subject-full-stop": [0, "never"],
      "subject-case": [0, "never"],
      "header-max-length": [0, "always", 72],
    },
    prompt: {
      messages: {
        skip: "回车跳过",
        max: "upper %d chars",
        min: "%d chars at least",
        emptyWarning: "内容不能为空",
        upperLimitWarning: "over limit",
        lowerLimitWarning: "below limit",
      },
      questions: {
        type: {
          description: "选择一个commit提交类型",
          enum: {
            feature: {
              description: "新增了什么功能",
              title: "Features",
              emoji: "✨",
            },
            fix: {
              description: "修补某功能的bug",
              title: "Bug Fixes",
              emoji: "🐛",
            },
            docs: {
              description: "仅文档新增/改动",
              title: "Documentation",
              emoji: "📚",
            },
            style: {
              description: "样式修改",
              title: "Styles",
              emoji: "💎",
            },
            refactor: {
              description: "重构某个功能",
              title: "Code Refactoring",
              emoji: "📦",
            },
            perf: {
              description: "性能, 体验优化",
              title: "Performance Improvements",
              emoji: "🚀",
            },
            test: {
              description: "测试某功能、新增测试用例、更新现有测试",
              title: "Tests",
              emoji: "🚨",
            },
            build: {
              description: "修改项目构建系统(例如 webpack，cli 的配置等)的提交",
              title: "Builds",
              emoji: "🛠",
            },
            ci: {
              description: "主要目的是修改项目继续集成流程",
              title: "Continuous Integrations",
              emoji: "⚙️",
            },
            chore: {
              description: "构建过程或辅助工具的变动",
              title: "Chores",
              emoji: "♻️",
            },
          },
        },
        scope: {
          description: "填写改动范围（修改了哪些组件、文件名）",
        },
        subject: {
          description: "大概描述改动内容",
        },
        body: {
          description: "可提供更详细的说明",
        },
        isBreaking: {
          description: "是否为破坏性修改？如接口删除、逻辑迁移、接口参数减少等",
        },
        isIssueAffected: {
          description: "改动修复了哪个问题？",
        },
      },
    },
  };
  

src/utils/shellCommand.ts
/*
 * @Date: 2022-03-14 22:28:20
 * @Author: handsome_anthony
 * @LastEditors: handsome_anthony
 * @Description: 导出一些shell命令
 */

import shell, { ShellString } from 'shelljs';

export async function closeAnySource(): Promise<void> {
  const command = 'sudo spctl --master-enable';

  try {
    console.log('运行命令：' + command);
    await shell.exec(command);
    console.log('已关闭任何来源');
  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message);
    }
  }
}

// 开启任何来源
export async function openAnySource(): Promise<void> {
  const command = 'sudo spctl --master-disable';

  try {
    console.log('运行命令：' + command);
    await shell.exec(command);
    console.log('已开启任何来源');
  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message);
    }
  }
}


// 安装oh my zsh
export async function installOhmyzsh(): Promise<void> {
  // 该命令可能需要终端能够访问github
  const installOhmyzshCommand = 'sh -c "$(curl -fsSL https://raw.github.com/ohmyzsh/ohmyzsh/master/tools/install.sh)"';
  try {
    // 首先判断ohmyzsh是否存在
    // 通过判断$ZSH 变量是否存在
    let zshPath = await shell.exec('echo $ZSH').stdout;

    // 过滤\n换行符号
    zshPath = zshPath.replace(/\n/, '');

    if (shell.test('-e', zshPath)) {
      console.log('oh-my-zsh已安装');
      return;
    }

    console.log('运行命令：' + installOhmyzshCommand);
    await shell.exec(installOhmyzshCommand);
    console.log('oh-my-zh安装完成');
    // 询问是否安装ohmyzsh的一些常用插件

  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message);
    }
  }
}


// 卸载oh my zsh 
export async function uninstallOhmyzsh(): Promise<void> {
  const uninstallOhmyzshCommand = 'uninstall_oh_my_zsh';

  try {
    // 首先判断ohmyzsh是否存在
    // 通过判断$ZSH 变量是否存在
    let zshPath = await shell.exec('echo $ZSH').stdout;

    // 过滤\n换行符号
    zshPath = zshPath.replace(/\n/, '');

    if (!shell.test('-e', zshPath)) {
      console.log('oh-my-zsh目录不存在');
      return;
    }

    // TODO 未知原因，暂时无法运行
    console.log('运行命令：' + uninstallOhmyzshCommand);
    const result = await shell.exec(uninstallOhmyzshCommand);

    if (result.stderr) {
      throw new Error(result.stderr);
    }

    console.log('oh-my-zsh卸载完成');

  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message);
    }
  }
}


// 将NVM的变量添加到zshrc中
export async function addNVMToZshrc(): Promise<void> {
  try {
    let nvmPath = await shell.exec('echo $NVM_DIR').stdout;
    // 过滤\n换行符
    nvmPath = nvmPath.replace(/\n/, '');

    if (nvmPath && shell.test('-e', nvmPath)) {
      console.log('NVM 全局变量已在zshrc中');
      return
    }

    // 将NVM变量添加到.zshrc文件底部
    const command = new shell.ShellString('\nexport NVM_DIR="$HOME/.nvm"\n[ -s "$NVM_DIR/nvm.sh" ] && \\. "$NVM_DIR/nvm.sh"  # This loads nvm\n[ -s "$NVM_DIR/bash_completion" ] && \\. "$NVM_DIR/bash_completion"  # This loads nvm bash_completion')
    command.toEnd('~/.zshrc');

    console.log('NVM变量添加成功');

  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message);
    }
  }
}


// 安装zsh-autosuggestions插件
export async function installZshAutosuggestions(): Promise<void> {
  try {
    // 要先判断zsh-autosuggestions文件夹是否已存在
    const zsh_autosuggestions_path = '~/.oh-my-zsh/custom/plugins/zsh-autosuggestions';

    // 如果目录不存在
    if (!shell.test('-e', zsh_autosuggestions_path)) {
      console.log('开始安装 zsh-autosuggestions');

      const installCommand = 'git clone https://github.91chi.fun/https://github.com/zsh-users/zsh-autosuggestions.git ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/zsh-autosuggestions';
      console.log('运行命令：' + installCommand);
      const result = await shell.exec(installCommand);
      // 如果有错误，则终止
      if (result.code !== 0) {
        console.error('安装zsh-autosuggestions遇到错误，请重试')
        return;
      }
      console.log('安装zsh-autosuggestions成功...');
    } else {
      console.log('监测到已安装zsh-autosuggestions');
    }

    // 安装完之后，要修改.zshrc文件，添加zsh-autosuggestions到plugins中
    const zshrcFilePath = '~/.zshrc';
    const { stdout: zshrcFile } = shell.cat(zshrcFilePath);

    if (!zshrcFile.includes('zsh-autosuggestions')) {
      console.log('更改~/.zshrc文件，写入zsh-autosuggestions...');
      // 这里加了一个\n，用于过滤注释中的plugins字段
      const reg = /\nplugins=\((.*)\)/g;
      // 这里用正则匹配plugins=() 括号中的内容，然后replace加入zsh-autosuggestions
      const newZshrcFile = zshrcFile.replace(reg, (_, p1) => {
        // p1 是括号中的内容
        return `\nplugins=(${p1 + ' ' + 'zsh-autosuggestions'})`;
      })
      
      // 将新的zshrc内容覆盖到~/.zshrc文件中
      new ShellString(newZshrcFile).to('~/.zshrc');
    } else {
      console.log('监测到zshrc中已经写入了zsh-autosuggestions');
    }

    console.log('完成');

  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message);
    }
  }
}

// 安装zsh-syntax-highlighting插件
export async function installZshSyntaxHighlighting(): Promise<void> {
  try {
    // 要先判断zsh-autosuggestions文件夹是否已存在
    const zsh_syntax_highlighting_path = '~/.oh-my-zsh/custom/plugins/zsh-syntax-highlighting';

    // 如果目录不存在
    if (!shell.test('-e', zsh_syntax_highlighting_path)) {
      console.log('开始安装 zsh-syntax-highlighting');

      const installCommand = 'git clone https://github.91chi.fun/https://github.com/zsh-users/zsh-syntax-highlighting.git ${ZSH_CUSTOM:-~/.oh-my-zsh/custom}/plugins/zsh-syntax-highlighting';
      console.log('运行命令：' + installCommand);
      const result = await shell.exec(installCommand);
      // 如果有错误，则终止
      if (result.code !== 0) {
        console.error('安装zsh-syntax-highlighting遇到错误，请重试')
        return;
      }
      console.log('安装zsh-syntax-highlighting成功...');
    } else {
      console.log('监测到已安装zsh-syntax-highlighting');
    }

    // 安装完之后，要修改.zshrc文件，添加zsh-syntax-highlighting到plugins中
    const zshrcFilePath = '~/.zshrc';
    const { stdout: zshrcFile } = shell.cat(zshrcFilePath);

    if (!zshrcFile.includes('zsh-syntax-highlighting')) {
      console.log('更改~/.zshrc文件，写入zsh-syntax-highlighting...');
      // 这里加了一个\n，用于过滤注释中的plugins字段
      const reg = /\nplugins=\((.*)\)/g;
      // 这里用正则匹配plugins=() 括号中的内容，然后replace加入zsh-syntax-highlighting
      const newZshrcFile = zshrcFile.replace(reg, (_, p1) => {
        // p1 是括号中的内容
        return `\nplugins=(${p1 + ' ' + 'zsh-syntax-highlighting'})`;
      })
      
      // 将新的zshrc内容覆盖到~/.zshrc文件中
      new ShellString(newZshrcFile).to('~/.zshrc');
    } else {
      console.log('监测到zshrc中已经写入了zsh-syntax-highlighting');
    }

    console.log('完成');

  } catch (error: unknown) {
    if (error instanceof Error) {
      throw new Error(error.message);
    }
  }
}